src/shims/tls.rs

```diff
@@ -328,21 +320,26 @@ trait EvalContextPrivExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         Ok(())
     }
 
-    /// Schedule the MacOS thread destructor of the thread local storage to be
-    /// executed.
-    fn schedule_macos_tls_dtor(&mut self) -> InterpResult<'tcx> {
+    /// Schedule the macOS thread local storage destructors to be executed.
+    fn schedule_macos_tls_dtors(&mut self) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
         let thread_id = this.active_thread();
-        if let Some((instance, data)) = this.machine.tls.macos_thread_dtors.remove(&thread_id) {
-            trace!("Running macos dtor {:?} on {:?} at {:?}", instance, data, thread_id);
-
-            this.call_function(
-                instance,
-                Abi::C { unwind: false },
-                &[data.into()],
-                None,
-                StackPopCleanup::Root { cleanup: true },
-            )?;
+        if let Some(dtors) = this.machine.tls.macos_thread_dtors.remove(&thread_id) {
+            // macOS runs TLS destructors in the reverse order of their registration.
+            // See https://github.com/apple-oss-distributions/dyld/blob/d552c40cd1de105f0ec95008e0e0c0972de43456/dyld/DyldRuntimeState.cpp#L2277
+            // FIXME: Old macOS versions had a use-after-free bug that triggered when
+            // _tlv_atexit was run inside a destructor. Maybe miri should catch that?
+            for (instance, data) in dtors.into_iter().rev() {
+                trace!("Running macos dtor {:?} on {:?} at {:?}", instance, data, thread_id);
+
+                this.call_function(
```

Comment:
No this is not right. `call_function` just pushes the stack frame, so what this does is push a stack frame for each of these functions and then run them all, making the stack look as if the 2nd dtor was called by the first.

There's a big surrounding interpreter loop here, and the next dtor needs to be started only once this dtor finishes running, when the stack is empty again.  See what we do for `WindowsDtors` for how that can be implemented.

Reply:
To give a bit of context here, each thread has a `TlsDtorsState` and each time that thread's stack is empty, the main loop calls `on_stack_empty` on it. If that returns `Pending`, the thread sticks around. This *may* also push a new stack frame, in which case that now gets executed, until the stack is empty and `on_stack_empty` is called again. This repeats until `on_stack_empty` returns `Ready`, at which point the thread is fully terminated and never scheduled again.